File: examples/reducing/pca.cpp
/**
 @cond
 ############################################################################
 # LGPL License                                                             #
 #                                                                          #
 # This file is part of the Machine Learning Framework.                     #
 # Copyright (c) 2010-2012, Philipp Kraus, <philipp.kraus@flashpixx.de>     #
 # This program is free software: you can redistribute it and/or modify     #
 # it under the terms of the GNU Lesser General Public License as           #
 # published by the Free Software Foundation, either version 3 of the       #
 # License, or (at your option) any later version.                          #
 #                                                                          #
 # This program is distributed in the hope that it will be useful,          #
 # but WITHOUT ANY WARRANTY; without even the implied warranty of           #
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            #
 # GNU Lesser General Public License for more details.                      #
 #                                                                          #
 # You should have received a copy of the GNU Lesser General Public License #
 # along with this program. If not, see <http://www.gnu.org/licenses/>.     #
 ############################################################################
 @endcond
 **/

#include <cstdlib>
#include <machinelearning.h>
#include <boost/numeric/ublas/matrix.hpp>
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/variables_map.hpp>
#include <boost/program_options/options_description.hpp>


namespace po    = boost::program_options;
namespace ublas = boost::numeric::ublas;
namespace dim   = machinelearning::dimensionreduce::nonsupervised;
namespace tools = machinelearning::tools;


/** main program
 * @param p_argc number of arguments
 * @param p_argv arguments
 **/
int main(int p_argc, char* p_argv[])
{
    #ifdef MACHINELEARNING_MULTILANGUAGE
    tools::language::bindings::bind();
    #endif

    // default values
    std::size_t l_dimension;
    std::string l_outpath;

    // create CML options with description
    po::options_description l_description("allowed options");
    l_description.add_options()
        ("help", "produce help message")
        ("infile", po::value<std::string>(), "input file")
        ("inpath", po::value<std::string>(), "input path of the datapoint within the input file")
        ("outfile", po::value<std::string>(), "output HDF5 file")
        ("outpath", po::value<std::string>(&l_outpath)->default_value("/pca"), "output path within the HDF5 file [default: /pca]")
        ("dimension", po::value<std::size_t>(&l_dimension)->default_value(3), "target dimension [default: 3]")
    ;

    po::variables_map l_map;
    po::positional_options_description l_input;
    po::store(po::command_line_parser(p_argc, p_argv).options(l_description).positional(l_input).run(), l_map);
    po::notify(l_map);

    if (l_map.count("help")) {
        std::cout << l_description << std::endl;
        return EXIT_SUCCESS;
    }

    if ( (!l_map.count("infile")) || (!l_map.count("inpath")) || (!l_map.count("outfile")) ) {
        std::cerr << "[--infile], [--inpath] and [--outfile] option must be set" << std::endl;
        return EXIT_FAILURE;
    }


    // read source hdf file
    tools::files::hdf source( l_map["infile"].as<std::string>() );

    // create pca object and map the data
    dim::pca<double> l_pca( l_dimension );
    const ublas::matrix<double> l_project = l_pca.map( source.readBlasMatrix<double>( l_map["inpath"].as<std::string>(), tools::files::hdf::NATIVE_DOUBLE) );


    // create file and write data to hdf
    tools::files::hdf target( l_map["outfile"].as<std::string>(), true);
    target.writeBlasMatrix<double>( l_outpath,  l_project, tools::files::hdf::NATIVE_DOUBLE );

    return EXIT_SUCCESS;
}

File: examples/java/reducing/pca.java
/** 
 @cond
 ############################################################################
 # LGPL License                                                             #
 #                                                                          #
 # This file is part of the Machine Learning Framework.                     #
 # Copyright (c) 2010-2012, Philipp Kraus, <philipp.kraus@flashpixx.de>     #
 # This program is free software: you can redistribute it and/or modify     #
 # it under the terms of the GNU Lesser General Public License as           #
 # published by the Free Software Foundation, either version 3 of the       #
 # License, or (at your option) any later version.                          #
 #                                                                          #
 # This program is distributed in the hope that it will be useful,          #
 # but WITHOUT ANY WARRANTY; without even the implied warranty of           #
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            #
 # GNU Lesser General Public License for more details.                      #
 #                                                                          #
 # You should have received a copy of the GNU Lesser General Public License #
 # along with this program. If not, see <http://www.gnu.org/licenses/>.     #
 ############################################################################
 @endcond
 **/


import machinelearning.dimensionreduce.nonsupervised.*;
import java.util.Random;


/** java testprogram for using PCA **/
public class pca {
    
    
    /** main method
     * @param p_args input arguments
     **/
    public static void main(String[] p_args)
    {
        // generates random datapoints
        Random l_rand = new Random();
        
        Double[][] l_data = new Double[15][8];
        for(int i=0; i < l_data.length; i++) {
            for (int j=0; j < l_data[i].length; j++) {
                l_data[i][j] = l_rand.nextDouble() * 500;
                System.out.print(l_data[i][j] + "\t");
            }
            System.out.println("");
        }

        // create PCA object
        PCA l_pca = new PCA(3);
        
        
        // maps the random data points
        Double[][] l_result = l_pca.map(l_data);
        System.out.println("\nproject data:");
        if (l_result == null)
            System.out.println("no data is returned");
        else
            for(int i=0; i < l_result.length; i++) {
                for(int j=0; j < l_result[i].length; ++j)
                    System.out.print(l_result[i][j] + "\t");
                System.out.println("");
            }
        
        // show PCA eigenvectors 
        Double[][] l_eig = l_pca.getProject();
        System.out.println("\neigenvectors:");
        if (l_eig == null)
            System.out.println("no data is returned");
        else
            for(int i=0; i < l_eig.length; i++) {
                for(int j=0; j < l_eig[i].length; ++j)
                    System.out.print(l_eig[i][j] + "\t");
                System.out.println("");
            }
        
        
        // delete manually the PCA object (call destructor), should be
        // do the finalizer, but there is no guarantee.
        l_pca.delete();
        
        
        // set objects to null for gc removing
        l_pca    = null;
        l_result = null;
        l_data   = null;
        l_eig    = null;        
        l_rand   = null;
    }
    
}